### PR TITLE
Adding chip applicability check to filter off unsupported architectures

### PR DESCRIPTION
**Pull request by jerryyin**
In this PR, I focus mostly on refactoring:
 - Added the new applicability filter on chip architectures.
 - Did a fair amount of refactoring to decouple applicability and parsing.
   - Future work, use a dedicated component instead of conv2d generator to judge if the config is valid
 - Fix `mlir-miopen-driver` such that it does rely on the command line options to initialize in all scenarios
 - Fix `mlir-miopen-lib` such that it doesn't try to parse in the lowering stages, and scoped parsing to only handle creation stage.

### CODE REVIEW

File: mlir/include/mlir/Dialect/MIOpen/Generator/Conv2dGenerator.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/MIOpen/Generator/Conv2dGenerator.h:80]: Thanks for catching the old method declarations

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp]: Might be better to rename to something like hasValidProblem() ?

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp:227]: Since we're here, might as well reject chips older than 0x900, which'll also catch a sscanf() failure?

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp]: Maybe this could be an `if(failed(parser.parseIsaName(...)))`?

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp
Comment (krzysz00) [inline mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp:380]: On the other hand, having a `status =` makes sense here.

File: mlir/test/mlir-miopen-lib/applicability/chip.mlir
Comment (krzysz00) [inline mlir/test/mlir-miopen-lib/applicability/chip.mlir]: I think the "expected failure" case can be handled with `not` , that is `// RUN: not mlir-miopen-lib-test --args ...`

That'll let us get rid of the FileCheck calls and the echo

File: mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp
Comment (krzysz00) [inline mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp]: I think this can be moved into the if, same for the other validity check. That is, `if(failed(conv2dGenerator.isValidChip())` as the test.

**Review by krzysz00 — COMMENTED** [review]
I've mainly got nitpicks, overall this all makes sense.

(I'm willing to approve even if you don't like my feedback)

**Comment by jerryyin** [comment]
> (I'm willing to approve even if you don't like my feedback)

Pleas wait till I addressed all feedback to give the approve. I indeed like mostly all the feedbacks you give. Thanks for catching these!

**Comment by krzysz00** [comment]
> I indeed like mostly all the feedbacks you give. Thanks for catching these!
Thank you!

My comment was more an indication that I didn't see anything _incorrect_ in here, just the sort of thing where people might disagree, and so "no, I don't like that change" wasn't going to be a barrier to an approve.

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp
Comment (jerryyin) [reply mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp]: I kept the old naming so that it isn't too generic. `hasValidProblem()` is almost as generic as the top level function `isApplicable()`

File: mlir/test/mlir-miopen-lib/applicability/chip.mlir
Comment (jerryyin) [reply mlir/test/mlir-miopen-lib/applicability/chip.mlir]: Thanks for letting me know of it, very convenient!

File: mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp
Comment (jerryyin) [reply mlir/lib/Dialect/MIOpen/Generator/Conv2dGenerator.cpp:227]: Agreed, addressed in 07e2608

File: mlir/test/mlir-miopen-lib/applicability/chip.mlir
Comment (krzysz00) [inline mlir/test/mlir-miopen-lib/applicability/chip.mlir:4]: I didn't know you can write `not` as `!` in lit tests

Cool

File: mlir/include/mlir/Dialect/MIOpen/Generator/Conv2dGenerator.h
Comment (krzysz00) [inline mlir/include/mlir/Dialect/MIOpen/Generator/Conv2dGenerator.h]: My one complaint is that it feels like the name should be `hasValidDimension()` (and also `hasValidChip()` below) since those make the code more grammatical for me. That is, "if is applicable" makes sense (and its out of context "if generator is applicable", but not "if generator is valid dimension", which I'd prefer to be "if generator has valid dimension"

**Review by krzysz00 — APPROVED** [review]
LGTM, aside from some minor nitpicking about names you can address or not

File: mlir/test/mlir-miopen-lib/applicability/chip.mlir
Comment (jerryyin) [reply mlir/test/mlir-miopen-lib/applicability/chip.mlir:4]: I thought ! is the only way for logical not, [here](https://bash.cyberciti.biz/guide/Logical_Not_!).

Comment (krzysz00) [reply mlir/test/mlir-miopen-lib/applicability/chip.mlir:4]: The llvm testsuite provides `not`, which probably has the side effect of being less platform-dependent https://llvm.org/docs/TestingGuide.html#other-features

File: mlir/include/mlir/Dialect/MIOpen/Generator/Conv2dGenerator.h
Comment (jerryyin) [reply mlir/include/mlir/Dialect/MIOpen/Generator/Conv2dGenerator.h]: Fair enough. I had the naming almost with the mindset of calling those functions from another component.

I feel like conv2d generator component does much more than it should have and it should be:
 - conv2d config/descriptor - the data structure that carry information to describe a conv2d
 - conv2d parser - that only parse conv2d configs to a descriptor
 - conv2d checker - that validate the parsed result
 - conv2d generator - that does only generation

Whereas now we sort of coupled everything together in one class.

File: mlir/test/mlir-miopen-lib/applicability/chip.mlir
Comment (jerryyin) [reply mlir/test/mlir-miopen-lib/applicability/chip.mlir:4]: Ha, didn't know that

File: mlir/include/mlir/Dialect/MIOpen/Generator/Conv2dGenerator.h
Comment (krzysz00) [reply mlir/include/mlir/Dialect/MIOpen/Generator/Conv2dGenerator.h]: Yeah, agreed

It doesn't help that we don't carry that descriptor around much - it's mainly used to create the conv2d op

**Pull request merged**
